server/jetstream_cluster.go

```diff
@@ -8005,196 +8005,225 @@ func (mset *stream) processClusteredInboundMsg(subject, reply string, hdr, msg [
 		return err
 	}
 
-	// Some header checks can be checked pre proposal. Most can not.
-	var msgId string
-	if len(hdr) > 0 {
-		// Since we encode header len as u16 make sure we do not exceed.
-		// Again this works if it goes through but better to be pre-emptive.
-		if len(hdr) > math.MaxUint16 {
-			err := fmt.Errorf("JetStream header size exceeds limits for '%s > %s'", jsa.acc().Name, mset.cfg.Name)
-			s.RateLimitWarnf("%s", err.Error())
+	if batchId := getBatchId(hdr); batchId != _EMPTY_ {
+		if !allowAtomicPublish {
+			err := NewJSAtomicPublishDisabledError()
 			if canRespond {
-				var resp = &JSPubAckResponse{PubAck: &PubAck{Stream: name}}
-				resp.Error = NewJSStreamHeaderExceedsMaximumError()
-				response, _ = json.Marshal(resp)
-				outq.send(newJSPubMsg(reply, _EMPTY_, _EMPTY_, nil, response, nil, 0))
+				b, _ := json.Marshal(&JSPubAckResponse{PubAck: &PubAck{Stream: name}, Error: err})
+				outq.send(newJSPubMsg(reply, _EMPTY_, _EMPTY_, nil, b, nil, 0))
 			}
 			return err
 		}
-		// Expected stream name can also be pre-checked.
-		if sname := getExpectedStream(hdr); sname != _EMPTY_ && sname != name {
+
+		batchSeq, exists := getBatchSequence(hdr)
+		if !exists {
+			err := NewJSAtomicPublishMissingSeqError()
 			if canRespond {
-				var resp = &JSPubAckResponse{PubAck: &PubAck{Stream: name}}
-				resp.PubAck = &PubAck{Stream: name}
-				resp.Error = NewJSStreamNotMatchError()
-				b, _ := json.Marshal(resp)
-				outq.sendMsg(reply, b)
+				b, _ := json.Marshal(&JSPubAckResponse{PubAck: &PubAck{Stream: name}, Error: err})
+				outq.send(newJSPubMsg(reply, _EMPTY_, _EMPTY_, nil, b, nil, 0))
 			}
-			return errStreamMismatch
+			return err
 		}
-		// TTL'd messages are rejected entirely if TTLs are not enabled on the stream, or if the TTL is invalid.
-		if ttl, err := getMessageTTL(hdr); !sourced && (ttl != 0 || err != nil) {
-			if !allowTTL {
-				if canRespond {
-					var resp = &JSPubAckResponse{PubAck: &PubAck{Stream: name}}
-					resp.Error = NewJSMessageTTLDisabledError()
-					b, _ := json.Marshal(resp)
-					outq.sendMsg(reply, b)
-				}
-				return errMsgTTLDisabled
-			} else if err != nil {
+		commit := len(sliceHeader(JSBatchCommit, hdr)) != 0
+
+		mset.mu.Lock()
+		if mset.batches == nil {
+			mset.batches = &batching{
+				group: make(map[string]*batchGroup, 1),
```

Comment:
Wonder if we should be making sure that this map can't grow infinitely, especially because maps never shrink.

Reply:
Indeed, the ADR mentions the following initial set of limits:
> The server will operate under limits to safeguard itself:
> * Each stream can only have 50 batches in flight at any time
> * Each server can only have 1000 batches in flight at any time
> * A batch that has not had traffic for 10 seconds will be abandoned
> * Each batch can have maximum 1000 messages

Have mentioned this in the current set of limitations as well, and is one of the things to be added assuming we agree on the API/ADR:
> There are no limits yet to the amount of batches, how large they are, how long the batch ID can be, when they time out and are abandoned. Timing out of batches is also not implemented yet.